spring-integration-test/src/main/java/org/springframework/integration/test/context/MockIntegrationContext.java

```diff
@@ -125,14 +126,41 @@ public void instead(String consumerEndpointId, MessageHandler mockMessageHandler
 	}
 
 	public void instead(String consumerEndpointId, MessageHandler mockMessageHandler, boolean autoStartup) {
-		instead(consumerEndpointId, mockMessageHandler, IntegrationConsumer.class, "handler", autoStartup);
+		Object endpoint = this.beanFactory.getBean(consumerEndpointId, IntegrationConsumer.class);
+		if (autoStartup && endpoint instanceof Lifecycle) {
+			((Lifecycle) endpoint).stop();
+		}
+		DirectFieldAccessor directFieldAccessor = new DirectFieldAccessor(endpoint);
+		Object targetMessageHandler = directFieldAccessor.getPropertyValue("handler");
+		this.beans.put(consumerEndpointId, targetMessageHandler);
+
 		if (mockMessageHandler instanceof MessageProducer) {
-			Object targetMessageHandler = this.beans.get(consumerEndpointId);
 			if (targetMessageHandler instanceof MessageProducer) {
 				MessageChannel outputChannel = TestUtils.getPropertyValue(targetMessageHandler, "outputChannel",
 						MessageChannel.class);
 				((MessageProducer) mockMessageHandler).setOutputChannel(outputChannel);
 			}
+			else {
+				if (mockMessageHandler instanceof MockMessageHandler) {
+					if (!TestUtils.getPropertyValue(mockMessageHandler, "hasReplies", Boolean.class)) {
+						mockMessageHandler = mockMessageHandler::handleMessage;
```

Comment:
Well, you've just widened my Java perception horizons. :smile: I've checked the code in IDEA and it highlighted it as a warning. Your intention appears to be more thoughtful than Jetbrains guys might suppose.
Agree with you on the matter.